### PR TITLE
fix(replica): fix replica reconnect handing

### PR DESCRIPTION
**Pull request by dranikpg**
* Makes FLUSHALL sync
~~* Removes preceding FLUSHALL to Replicaof~~
* Refactors ReplicaOf
* Fixes bug in Replica with reply

### CODE REVIEW

File: src/server/db_slice.cc [hunk 1]
```diff
@@ -633,12 +633,11 @@ void DbSlice::FlushDb(DbIndex db_ind) {
...
-    for (auto& db : all_dbs) {
-      db.reset();
-    }
-    mi_heap_collect(ServerState::tlocal()->data_heap(), true);
-  }).Detach();
+  // Explicitly drop reference counted pointers in place.
+  // If snapshotting is currently in progress, they will keep alive until it finishes.
+  for (auto& db : all_dbs)
+    db.reset();
+  mi_heap_collect(ServerState::tlocal()->data_heap(), true);
```
Comment (dranikpg) [inline src/server/db_slice.cc:640]: all I changed to make it synchronous

File: src/server/replica.cc [hunk 2]
```diff
@@ -81,22 +81,23 @@ error_code Replica::Start(ConnectionContext* cntx) {
...
     }
-    return {};
+    return ec;
   };
 
+  // 0. Set basic error handler that is reponsible for cleaning up on errors.
+  // Can return an error only if replication was cancelled immediately.
+  auto err = cntx_.SwitchErrorHandler([this](const auto& ge) { this->DefaultErrorHandler(ge); });
+  RETURN_ON_ERR(check_connection_error(err, "replication cancelled"));
+
```
Comment (dranikpg) [inline src/server/replica.cc:100]: this was the bug, there is a 0.00001% chance we're cancelled right after the start while setting up the context, it does an atomic swap-err-handler-or-return-err and we just returned in that case

because it's 100% a cancellation, I just let  check_connection_error handle it instead of replying manually

File: src/server/server_family.cc [hunk 3]
```diff
@@ -1577,11 +1582,10 @@ void ServerFamily::Info(CmdArgList args, ConnectionContext* cntx) {
...
 
-      // it's safe to access replica_ because replica_ is created before etl.is_master set to
-      // false and cleared after etl.is_master is set to true. And since the code here that checks
-      // for is_master and copies shared_ptr is atomic, it1 should be correct.
-      auto replica_ptr = replica_;
-      Replica::Info rinfo = replica_ptr->GetInfo();
+      // The replica pointer can still be mutated even while master=true,
+      // we don't want to drop the replica object in this fiber
+      unique_lock lk{replicaof_mu_};
+      Replica::Info rinfo = replica_->GetInfo();
```
Comment (dranikpg) [inline src/server/server_family.cc:1588]: I did not investigate 😢 But when I dropped it here all kinds of fiber / eventcount asserts failed badly

File: src/server/server_family.cc [hunk 4]
```diff
@@ -1820,33 +1817,20 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
...
-    }
-    bool is_master = !replica_;
-    pool.AwaitFiberOnAll(
-        [&](util::ProactorBase* pb) { ServerState::tlocal()->is_master = is_master; });
-  } else {
-    new_replica->Stop();
+  if (ec && replica_ == new_replica) {
+    service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
+    SetMasterFlagOnAllThreads(false);
+    replica_.reset();
```
Comment (dranikpg) [inline src/server/server_family.cc:1826]: If `replica == new_replica`, then no changed to the replica configuration were made. Then how come that `is_master = true` ? the first branch does no useful work for the case `ec = false`

Second, if the replica changed, then at least one time ReplicaOf was called, meaning the one following after us saw that replica was assgined and called replica_->Stop, meaning we should be stopped now 

There's surely nothing wrong with this code, but those big redundant constructs make it harder to read and understand this function

File: src/server/server_family.cc [hunk 5]
```diff
@@ -1749,29 +1753,20 @@ void ServerFamily::Hello(CmdArgList args, ConnectionContext* cntx) {
...
-  // replication, it will acquire the lock, call Stop() on the old replica_ and wait for Stop() to
-  // complete. So Replica::Stop() must
-  // 1. Be very responsive, as it is called while holding the lock.
-  // 2. Leave the DB in a consistent state after it is done.
-  // We have a relatively involved state machine inside Replica itself which handels cancellation
-  // with those requirements.
-  VLOG(2) << "Acquire replica lock";
+  // Wait until all transactions finished and prevent any other from executing
+  cntx->transaction->Schedule();
+  cntx->transaction->Execute([](auto*, auto*) { return OpStatus::OK; }, false);
```
Comment (dranikpg) [inline src/server/server_family.cc:1760]: I'll transfer the comment from above. Technically it was wrong, because for example `We lock to protect global state changes that we perform during the replication setup: the replica_ pointer, GlobalState, and the DB itself`. We did _not_ lock the DB. We do now... But is there actually any point in it? We set the global state, so we have a clear boundary... 🤔

**Comment by dranikpg** [comment]
There may be a few fixes, comments & clarifications left, but let's first agree on what features to keep

File: src/server/server_family.cc (hunk 3 above)
Comment (adiholden) [reply src/server/server_family.cc:1588]: I believe this is because of ABSL_GUARDED_BY
std::shared_ptr<Replica> replica_ ABSL_GUARDED_BY(replicaof_mu_);
@chakaz any idea why Vlad sees now this asserts fail but we didnt see it before?

File: src/server/server_family.cc [hunk 6]
```diff
@@ -1820,33 +1825,20 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
...
   }
 }
 
 void ServerFamily::ReplicaOf(CmdArgList args, ConnectionContext* cntx) {
   string_view host = ArgS(args, 0);
   string_view port = ArgS(args, 1);
 
-  // don't flush if input is NO ONE
-  if (!IsReplicatingNoOne(host, port))
-    Drakarys(cntx->transaction, DbSlice::kDbAll);
```
Comment (adiholden) [inline src/server/server_family.cc:1848]: you removed the call to Drakarys here but I dont see it inside ReplicaOfInternal

File: src/server/server_family.cc (hunk 3 above)
Comment (chakaz) [reply src/server/server_family.cc:1588]: I'm slightly out of context, but when you say asserts you mean at runtime, right?
`ABSL_GUARDED_BY` is a compile-time-only feature, so it shouldn't affect runtime

File: src/server/server_family.cc [hunk 7]
```diff
@@ -1783,34 +1786,36 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
...
+      new_state != GlobalState::LOADING) {
+    LOG(WARNING) << GlobalStateName(new_state) << " in progress, ignored";
+    (*cntx)->SendError("Invalid state");
+    return;
+  }
 
-  if (replica_) {
-    replica_->Stop();  // NOTE: consider introducing update API flow.
-  } else {
-    // TODO: to disconnect all the blocked clients (pubsub, blpop etc)
```
Comment (adiholden) [inline src/server/server_family.cc:1797]: why did you remove the TODO comment?

File: src/server/server_family.cc (hunk 6 above)
Comment (dranikpg) [reply src/server/server_family.cc:1848]: that's what I asked above, we already have a flushall when we start full sync, do we need another one here?

File: src/server/server_family.cc (hunk 7 above)
Comment (dranikpg) [reply src/server/server_family.cc:1797]: Ah yes, will keep it

File: src/server/server_family.cc [hunk 8]
```diff
@@ -1820,33 +1825,20 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
...
-      replica_.reset();
-    }
-    bool is_master = !replica_;
-    pool.AwaitFiberOnAll(
-        [&](util::ProactorBase* pb) { ServerState::tlocal()->is_master = is_master; });
-  } else {
-    new_replica->Stop();
+  if (ec && replica_ == new_replica) {
+    service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
+    SetMasterFlagOnAllThreads(false);
```
Comment (adiholden) [inline src/server/server_family.cc:1833]: I am confused. Shouldnt we set the flag to master = true if the replication attempt failed ?
SetMasterFlagOnAllThreads(true)

File: src/server/server_family.cc (hunk 6 above)
Comment (adiholden) [reply src/server/server_family.cc:1848]: We do flushall on right before fullsync starts, what if we start replicaof and some of the stages before fullsync starts fails, should we flush the db in this case? If the answer is no than we can remove this flushall , otherwise we should keep it

File: src/server/server_family.cc (hunk 8 above)
Comment (dranikpg) [reply src/server/server_family.cc]: Fixed 😳😳😳

File: src/server/server_family.cc (hunk 6 above)
Comment (dranikpg) [reply src/server/server_family.cc:1848]: Ok, I'll keep it

File: src/server/server_family.cc [hunk 9]
```diff
@@ -1757,16 +1757,6 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
 
   // Wait until all transactions finished and prevent any other from executing.
   // The --replicaof flag executed on startup has no corresponding transaction.
-  if (auto* tx = cntx->transaction; tx) {
```
Comment (adiholden) [inline src/server/server_family.cc]: Why did you remove this ?

**Comment by kostasrim** [comment]
I will review this tomorrow extensively -- more eyes are always better :)

File: src/server/server_family.cc (hunk 9 above)
Comment (dranikpg) [reply src/server/server_family.cc]: Because I decided not do this functional change. I got no clear answer. The comment above it excessive (fixed)

**Review by adiholden — APPROVED** [review]

**Comment by dranikpg** [comment]
Kostas, sorry, you didn't make it in time 🙂
